### PR TITLE
Issue-6473: Remove non-unique caches from current list

### PR DESCRIPTION
**Pull request by FUSSEL4**
### Base of the PR
The solution provided here solves the [Issue-6473](https://github.com/cgeo/cgeo/issues/6473).

### Scope of the PR
The scope of the PR is to provide a functionality that removes all caches from the current list that are stored on other lists.

### Considerations
* Background of the feature request: After importing a gpx created by a pocket query it was hard to detect which cache was new, i.e. was not yet stored on the device (in any list).
* The method provided here solves the problem by removing all caches that are stored on more than the current list. Although removing caches the method is not _destructive_, because the removed caches are on other lists.
* The method takes filtering into account.
* An UNDO of the action is possible.
* The functionality itself is extracted to `RemoveNotUniqueHelper` to ensure testability.

### Testing
* Unit tests provided. See class `RemoveNotUniqueHelperTest`.
* Functionality was tested manually (Samsung Galaxy S3, S6).

### CODE REVIEW

File: main/res/values-de/strings.xml [hunk 1]
```diff
@@ -361,6 +361,7 @@
   <string name="caches_clear_offlinelogs_message">Möchtest du die Offline-Logs löschen?</string>
   <string name="caches_clear_offlinelogs_progress">Lösche Offline-Logs</string>
   <string name="caches_make_list_unique">Caches nur auf diese Liste</string>
+  <string name="caches_delete_not_singles">Entferne Nicht-Single-Caches</string>
```
Comment (samueltardieu) [inline main/res/values-de/strings.xml]: Please only submit changes to `main/res/values`, translations will be created in Crowdin.

File: main/src/cgeo/geocaching/CacheListActivity.java [hunk 2]
```diff
@@ -101,6 +102,7 @@
 import android.widget.AdapterView.AdapterContextMenuInfo;
 import android.widget.ListView;
 import android.widget.TextView;
+import android.widget.Toast;
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/CacheListActivity.java]: `Toast` is not used anywhere in the rest of the changes to this file (while all the changes are seen together). It must be wrong.

File: main/res/values-de/strings.xml (hunk 1 above)
Comment (FUSSEL4) [inline main/res/values-de/strings.xml]: OK, I'll remove all other translations.

File: main/src/cgeo/geocaching/CacheListActivity.java (hunk 2 above)
Comment (FUSSEL4) [inline main/src/cgeo/geocaching/CacheListActivity.java]: import removed

**Comment by FUSSEL4** [comment]
@samueltardieu Please, have another look at my PR. Thank you.

File: main/src/cgeo/geocaching/CacheListActivity.java [hunk 3]
```diff
@@ -1,5 +1,38 @@
...
+import android.view.ContextMenu;
+import android.view.ContextMenu.ContextMenuInfo;
+import android.view.KeyEvent;
+import android.view.Menu;
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.AdapterView.AdapterContextMenuInfo;
+import android.widget.ListView;
+import android.widget.TextView;
+import butterknife.ButterKnife;
```
Comment (UniQP) [inline main/src/cgeo/geocaching/CacheListActivity.java]: Here, you changed the imports in an undesired manner: https://github.com/cgeo/cgeo/wiki/Coding-Conventions

File: main/src/cgeo/geocaching/command/RemoveNotUniqueHelper.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package cgeo.geocaching.command;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import cgeo.geocaching.models.Geocache;
```
Comment (UniQP) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueHelper.java]: Please sort the imports as stated in https://github.com/cgeo/cgeo/wiki/Coding-Conventions

File: main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import cgeo.geocaching.enumerations.LoadFlags;
+import cgeo.geocaching.models.Geocache;
+import cgeo.geocaching.storage.DataStore;
+
+/**
+ * Removes caches from the current list which are already stored on other lists.
+ */
+public abstract class RemoveNotUniqueCommand extends AbstractCommand {
+
+  private final SearchResult searchResult;
```
Comment (UniQP) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java]: Please use four spaces for indentation.

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
+package cgeo.geocaching.command;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
+import junit.framework.TestCase;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import cgeo.geocaching.models.Geocache;
```
Comment (UniQP) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: Please sort the imports as stated in https://github.com/cgeo/cgeo/wiki/Coding-Conventions

File: main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Set;
+
+import cgeo.geocaching.R;
+import cgeo.geocaching.SearchResult;
+import cgeo.geocaching.enumerations.LoadFlags;
+import cgeo.geocaching.models.Geocache;
+import cgeo.geocaching.storage.DataStore;
```
Comment (UniQP) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java]: Please sort the imports as stated in https://github.com/cgeo/cgeo/wiki/Coding-Conventions

File: main/res/values/strings.xml [hunk 8]
```diff
@@ -1551,6 +1552,7 @@
     <string name="command_rename_list_result">Renamed list</string>
     <string name="command_delete_list_result">Deleted list</string>
     <string name="command_unique_list_result">Deleted caches from other lists</string>
+    <string name="command_remove_not_unique_result">Deleted caches from current list that are on other lists</string>
```
Comment (UniQP) [inline main/res/values/strings.xml]: Wording: `Removed caches that are on other lists from the current list`

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        lists3.add(LIST_ID3);
+        // WHEN
+        final Set<Geocache> toBeRemoved = RemoveNotUniqueHelper.removeNonUniqueCaches(geoCaches);
+        // THEN
+        assertThat(toBeRemoved.size()).isEqualTo(3);
+        assertThat(toBeRemoved.contains(gc1)).isTrue();
+        assertThat(toBeRemoved.contains(gc2)).isTrue();
+        assertThat(toBeRemoved.contains(gc3)).isTrue();
+    }
+    public void testShouldRemoveNoCache() throws Exception {
```
Comment (UniQP) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: Please add an empty line between the two methods.

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        geoCaches.add(gc1);
+        geoCaches.add(gc2);
+        geoCaches.add(gc3);
+
+        lists1 = gc1.getLists();
+        lists2 = gc2.getLists();
+        lists3 = gc3.getLists();
+    }
+
+    public void testRemoveNonUniqueCaches() throws Exception {
```
Comment (UniQP) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: Is the `throws Exception` needed for the test methods?

File: main/res/values/strings.xml [hunk 11]
```diff
@@ -377,6 +377,7 @@
     <string name="caches_clear_offlinelogs_message">Do you want to clear the offline logs?</string>
     <string name="caches_clear_offlinelogs_progress">Clearing offline logs</string>
     <string name="caches_make_list_unique">Make list unique</string>
+    <string name="caches_delete_not_singles">Remove not unique caches</string>
```
Comment (FUSSEL4) [inline main/res/values/strings.xml]: Changed.

File: main/res/values/strings.xml (hunk 8 above)
Comment (FUSSEL4) [inline main/res/values/strings.xml]: I changed it to `Removed caches from current list that are on other lists`.

File: main/src/cgeo/geocaching/CacheListActivity.java (hunk 3 above)
Comment (FUSSEL4) [inline main/src/cgeo/geocaching/CacheListActivity.java]: Changed to original order

File: main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java (hunk 7 above)
Comment (FUSSEL4) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java]: changed.

File: main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java (hunk 5 above)
Comment (FUSSEL4) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueCommand.java]: corrected

File: main/src/cgeo/geocaching/command/RemoveNotUniqueHelper.java (hunk 4 above)
Comment (FUSSEL4) [inline main/src/cgeo/geocaching/command/RemoveNotUniqueHelper.java]: corrected.

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java (hunk 6 above)
Comment (FUSSEL4) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: changed

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java (hunk 10 above)
Comment (FUSSEL4) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: removed

File: tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java (hunk 9 above)
Comment (FUSSEL4) [inline tests/src/cgeo/geocaching/command/RemoveNotUniqueHelperTest.java]: Empty line inserted

**Comment by FUSSEL4** [comment]
@UniQP Thanks for your review. I changed my implementation accordingly.

File: main/res/values/strings.xml (hunk 8 above)
Comment (UniQP) [inline main/res/values/strings.xml]: From "The Elements of Style":
>  The relative pronoun should come, in most instances, immediately after its antecedent.

Comment (UniQP) [inline main/res/values/strings.xml]: I still prefer `the current list`.

**Comment by Lineflyer** [comment]
I think also this PR needs to be redone.

**Comment by FUSSEL4** [comment]
I'll make a new PR out of thids changes.

**Pull request closed without merging**